wheel/python/tests/test_program.py

```diff
@@ -0,0 +1,495 @@
+    prog = Program.fromhex("ff10ff02ff0580")  # `(+ 2 5)`
+    args = Program.fromhex("01")  # `1`
+    assert plus.uncurry() == (prog, [args])
+
+
+def test_uncurry_top_level_garbage():
+    # there's garbage at the end of the top-level list
+    # `(a (q . 1) (c (q . 1) (q . 1)) (q . 0x1337))`
+    plus = Program.fromhex("ff02ffff0101ffff04ffff0101ffff010180ffff0182133780")
+    assert plus.uncurry() == (plus, None)
+
+
+def test_uncurry_not_pair():
+    # the second item in the list is expected to be a pair, with a qoute
+    # `(a 1 (c (q . 1) (q . 1)))`
+    plus = Program.fromhex("ff02ff01ffff04ffff0101ffff01018080")
+    assert plus.uncurry() == (plus, None)
+
+
+def test_uncurry_args_garbage():
+    # there's garbage at the end of the args list
+    # `(a (q . 1) (c (q . 1) (q . 1) (q . 4919)))`
+    plus = Program.fromhex("ff02ffff0101ffff04ffff0101ffff0101ffff018213378080")
+    assert plus.uncurry() == (plus, None)
+
+
+class SimpleStorage(CLVMStorage):
+    """
+    A simple implementation of `CLVMStorage`.
+    """
+
+    atom: Optional[bytes]
+
+    def __init__(self, atom, pair):
+        self.atom = atom
+        self._pair = pair
+
+    @property
+    def pair(self) -> Optional[Tuple["CLVMStorage", "CLVMStorage"]]:
+        return self._pair
+
+
+class Uncachable(SimpleStorage):
+    """
+    This object does not allow `_cached_sha256_treehash` or `_cached_serialization`
+    to be stored.
+    """
+
+    def get_th(self):
+        return None
+
+    def set_th(self, v):
+        raise AttributeError("can't set property")
+
+    _cached_sha256_treehash = property(get_th, set_th)
+    _cached_serialization = property(get_th, set_th)
+
+
+def convert_atom_to_bytes(castable: Any) -> Optional[bytes]:
+    return Program.to(castable).atom
+
+
+def validate_program(program):
+    validate_stack = [program]
+    while validate_stack:
+        v = validate_stack.pop()
+        assert isinstance(v, Program)
+        if v.pair:
+            assert isinstance(v.pair, tuple)
+            v1, v2 = v.pair
+            assert is_clvm_storage(v1)
+            assert is_clvm_storage(v2)
+            s1, s2 = v.pair
+            validate_stack.append(s1)
+            validate_stack.append(s2)
+        else:
+            assert isinstance(v.atom, bytes)
+
+
+def print_leaves(tree: Program) -> str:
+    a = tree.atom
+    if a is not None:
+        if len(a) == 0:
+            return "() "
+        return "%d " % a[0]
+
+    ret = ""
+    pair = tree.pair
+    assert pair is not None
+    for i in pair:
+        ret += print_leaves(i)
+
+    return ret
+
+
+def print_tree(tree: Program) -> str:
+    a = tree.atom
+    if a is not None:
+        if len(a) == 0:
+            return "() "
+        return "%d " % a[0]
+
+    pair = tree.pair
+    assert pair is not None
+    ret = "("
+    for i in pair:
+        ret += print_tree(i)
+    ret += ")"
+    return ret
+
+
+class ProgramTest(TestCase):
```

Comment:
I still have a soft spot for `unittest` since it's built-in and doesn't require `pytest`, although I suppose you're right that there's not really much benefit here, and it could possibly increase confusion to someone who hasn't lived through the `unittest` -> `pytest` ecosystem shift.

Reply:
Turns out that making this change would require rewriting both `self.assertEqual` and `self.assertRaises`, which I'd prefer not to do because I'm worried there might be something subtle going on in them. Probably not, but I think I'm going to just leave this as-is for now.

This is inherited code stolen from `chia-blockchain` or `clvm`, so it could be quite old by this point. I think it's best not to poke the beast too hard here.